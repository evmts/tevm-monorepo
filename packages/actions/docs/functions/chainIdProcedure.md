[**@tevm/actions**](../README.md)

***

[@tevm/actions](../globals.md) / chainIdProcedure

# Function: chainIdProcedure()

> **chainIdProcedure**(`baseClient`): [`EthChainIdJsonRpcProcedure`](../type-aliases/EthChainIdJsonRpcProcedure.md)

Defined in: [packages/actions/src/eth/chainIdProcedure.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/eth/chainIdProcedure.js#L8)

## Parameters

### baseClient

`TevmNode`\<`"fork"` \| `"normal"`, \{ \}\>

## Returns

[`EthChainIdJsonRpcProcedure`](../type-aliases/EthChainIdJsonRpcProcedure.md)
