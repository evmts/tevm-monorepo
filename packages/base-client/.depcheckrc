ignores: [
  "@ethereumjs/evm"
]
skip-missing: true
