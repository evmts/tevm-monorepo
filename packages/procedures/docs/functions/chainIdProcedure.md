**@tevm/procedures** ∙ [README](../README.md) ∙ [API](../API.md)

***

[API](../API.md) > chainIdProcedure

# Function: chainIdProcedure()

> **chainIdProcedure**(`chainId`): `EthChainIdJsonRpcProcedure`

## Parameters

▪ **chainId**: `number`

## Returns

## Source

[procedures/src/eth/chainIdProcedure.js:8](https://github.com/evmts/tevm-monorepo/blob/main/packages/procedures/src/eth/chainIdProcedure.js#L8)

***
Generated using [typedoc-plugin-markdown](https://www.npmjs.com/package/typedoc-plugin-markdown) and [TypeDoc](https://typedoc.org/)
