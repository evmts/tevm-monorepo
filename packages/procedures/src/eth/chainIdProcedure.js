import { chainIdHandler } from '@tevm/actions'
import { numberToHex } from '@tevm/utils'

/**
 * @param {() => Promise<number>} getChainId
 * @returns {import('@tevm/procedures-types').EthChainIdJsonRpcProcedure}
 */
export const chainIdProcedure = (getChainId) => async (req) => ({
	...(req.id ? { id: req.id } : {}),
	jsonrpc: '2.0',
	method: req.method,
	result: await chainIdHandler({ getChainId })({}).then(numberToHex),
})
