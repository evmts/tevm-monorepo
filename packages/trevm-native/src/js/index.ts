import { createLogger } from '@tevm/logger';
import { InternalError } from '@tevm/errors';
import type { StateManager } from '@tevm/state';
import type { Common } from '@tevm/common';
import type { Chain } from '@tevm/blockchain';

// Import native module
import { init_trevm, TrevmWrapper } from '../../dist/index.node';

export interface TrevmEvmOptions {
  common: Common;
  stateManager: StateManager;
  blockchain: Chain;
  customPrecompiles?: any[];
  profiler?: boolean;
  allowUnlimitedContractSize?: boolean;
  loggingLevel?: string;
}

export interface TrevmRunCallOptions {
  caller?: string;
  to?: string;
  data?: string;
  value?: bigint;
  gasLimit?: bigint;
  skipBalance?: boolean;
}

export interface TrevmRunCallResult {
  returnValue: string;
  executionGasUsed: string;
  error?: string;
}

export interface TrevmEvm {
  runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult>;
}

/**
 * Creates a trevm-based EVM instance
 * @param options EVM options
 * @returns A trevm EVM instance
 */
export const createTrevm = async ({
  common,
  stateManager,
  blockchain,
  customPrecompiles,
  profiler,
  allowUnlimitedContractSize,
  loggingLevel,
}: TrevmEvmOptions): Promise<TrevmEvm> => {
  const logger = createLogger({
    name: '@tevm/trevm',
    level: loggingLevel ?? 'warn',
  });
  
  logger.debug({
    allowUnlimitedContractSize,
    profiler,
    customPrecompiles: customPrecompiles?.map(c => c.address.toString()),
  });

  try {
    // Initialize the native trevm module with callbacks to the StateManager
    const nativeTrevm = await init_trevm({
      chainId: common.chainId(),
      allowUnlimitedContractSize: allowUnlimitedContractSize ?? false,
      stateCallbacks: {
        getAccount: async (address: string) => {
          try {
            const account = await stateManager.getAccount(address);
            if (!account) return null;
            
            return {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? `0x${Buffer.from(account.codeHash).toString('hex')}` : undefined,
              code: account.code ? `0x${Buffer.from(account.code).toString('hex')}` : undefined,
            };
          } catch (error) {
            logger.error(`Error in getAccount for ${address}:`, error);
            return null;
          }
        },
        putAccount: async (address: string, account: any) => {
          try {
            await stateManager.putAccount(address, {
              nonce: account.nonce,
              balance: account.balance,
              codeHash: account.codeHash ? Buffer.from(account.codeHash.slice(2), 'hex') : undefined,
              code: account.code ? Buffer.from(account.code.slice(2), 'hex') : undefined,
            });
            return true;
          } catch (error) {
            logger.error(`Error in putAccount for ${address}:`, error);
            return false;
          }
        },
        getStorage: async (address: string, key: string) => {
          try {
            const value = await stateManager.getContractStorage(address, key);
            return value ? `0x${Buffer.from(value).toString('hex')}` : '0x0';
          } catch (error) {
            logger.error(`Error in getStorage for ${address} key ${key}:`, error);
            return '0x0';
          }
        },
        putStorage: async (address: string, key: string, value: string) => {
          try {
            await stateManager.putContractStorage(
              address,
              key,
              Buffer.from(value.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putStorage for ${address} key ${key}:`, error);
            return false;
          }
        },
        getCode: async (address: string) => {
          try {
            const code = await stateManager.getContractCode(address);
            return code ? `0x${Buffer.from(code).toString('hex')}` : '0x';
          } catch (error) {
            logger.error(`Error in getCode for ${address}:`, error);
            return '0x';
          }
        },
        putCode: async (address: string, code: string) => {
          try {
            await stateManager.putContractCode(
              address,
              Buffer.from(code.slice(2), 'hex')
            );
            return true;
          } catch (error) {
            logger.error(`Error in putCode for ${address}:`, error);
            return false;
          }
        },
        getBlockHash: async (blockNumber: number) => {
          try {
            const block = await blockchain.getBlock(BigInt(blockNumber));
            return block ? `0x${Buffer.from(block.header.hash()).toString('hex')}` : null;
          } catch (error) {
            logger.error(`Error in getBlockHash for block ${blockNumber}:`, error);
            return null;
          }
        }
      }
    });

    // Return a JavaScript wrapper around the native trevm that matches the Evm interface
    return {
      async runCall(opts: TrevmRunCallOptions): Promise<TrevmRunCallResult> {
        try {
          return await nativeTrevm.run_call({
            caller: opts.caller,
            to: opts.to,
            data: opts.data,
            value: opts.value,
            gas_limit: opts.gasLimit,
            skip_balance: opts.skipBalance,
          });
        } catch (error) {
          logger.error('Error in runCall:', error);
          throw new InternalError(`Trevm execution error: ${error}`);
        }
      }
    };
  } catch (error) {
    logger.error('Error initializing trevm:', error);
    throw new InternalError(`Failed to initialize trevm: ${error}`);
  }
};