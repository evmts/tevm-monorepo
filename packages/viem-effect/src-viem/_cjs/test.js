'use strict'
Object.defineProperty(exports, '__esModule', { value: true })
exports.stopImpersonatingAccount =
	exports.setRpcUrl =
	exports.snapshot =
	exports.setStorageAt =
	exports.setNonce =
	exports.setNextBlockTimestamp =
	exports.setNextBlockBaseFeePerGas =
	exports.setMinGasPrice =
	exports.setLoggingEnabled =
	exports.setIntervalMining =
	exports.setCoinbase =
	exports.setCode =
	exports.setBlockTimestampInterval =
	exports.setBlockGasLimit =
	exports.setBalance =
	exports.setAutomine =
	exports.sendUnsignedTransaction =
	exports.revert =
	exports.reset =
	exports.removeBlockTimestampInterval =
	exports.mine =
	exports.inspectTxpool =
	exports.increaseTime =
	exports.impersonateAccount =
	exports.getTxpoolStatus =
	exports.getTxpoolContent =
	exports.getAutomine =
	exports.dropTransaction =
		void 0
const dropTransaction_js_1 = require('./actions/test/dropTransaction.js')
Object.defineProperty(exports, 'dropTransaction', {
	enumerable: true,
	get: function () {
		return dropTransaction_js_1.dropTransaction
	},
})
const getAutomine_js_1 = require('./actions/test/getAutomine.js')
Object.defineProperty(exports, 'getAutomine', {
	enumerable: true,
	get: function () {
		return getAutomine_js_1.getAutomine
	},
})
const getTxpoolContent_js_1 = require('./actions/test/getTxpoolContent.js')
Object.defineProperty(exports, 'getTxpoolContent', {
	enumerable: true,
	get: function () {
		return getTxpoolContent_js_1.getTxpoolContent
	},
})
const getTxpoolStatus_js_1 = require('./actions/test/getTxpoolStatus.js')
Object.defineProperty(exports, 'getTxpoolStatus', {
	enumerable: true,
	get: function () {
		return getTxpoolStatus_js_1.getTxpoolStatus
	},
})
const impersonateAccount_js_1 = require('./actions/test/impersonateAccount.js')
Object.defineProperty(exports, 'impersonateAccount', {
	enumerable: true,
	get: function () {
		return impersonateAccount_js_1.impersonateAccount
	},
})
const increaseTime_js_1 = require('./actions/test/increaseTime.js')
Object.defineProperty(exports, 'increaseTime', {
	enumerable: true,
	get: function () {
		return increaseTime_js_1.increaseTime
	},
})
const inspectTxpool_js_1 = require('./actions/test/inspectTxpool.js')
Object.defineProperty(exports, 'inspectTxpool', {
	enumerable: true,
	get: function () {
		return inspectTxpool_js_1.inspectTxpool
	},
})
const mine_js_1 = require('./actions/test/mine.js')
Object.defineProperty(exports, 'mine', {
	enumerable: true,
	get: function () {
		return mine_js_1.mine
	},
})
const removeBlockTimestampInterval_js_1 = require('./actions/test/removeBlockTimestampInterval.js')
Object.defineProperty(exports, 'removeBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return removeBlockTimestampInterval_js_1.removeBlockTimestampInterval
	},
})
const reset_js_1 = require('./actions/test/reset.js')
Object.defineProperty(exports, 'reset', {
	enumerable: true,
	get: function () {
		return reset_js_1.reset
	},
})
const revert_js_1 = require('./actions/test/revert.js')
Object.defineProperty(exports, 'revert', {
	enumerable: true,
	get: function () {
		return revert_js_1.revert
	},
})
const sendUnsignedTransaction_js_1 = require('./actions/test/sendUnsignedTransaction.js')
Object.defineProperty(exports, 'sendUnsignedTransaction', {
	enumerable: true,
	get: function () {
		return sendUnsignedTransaction_js_1.sendUnsignedTransaction
	},
})
const setAutomine_js_1 = require('./actions/test/setAutomine.js')
Object.defineProperty(exports, 'setAutomine', {
	enumerable: true,
	get: function () {
		return setAutomine_js_1.setAutomine
	},
})
const setBalance_js_1 = require('./actions/test/setBalance.js')
Object.defineProperty(exports, 'setBalance', {
	enumerable: true,
	get: function () {
		return setBalance_js_1.setBalance
	},
})
const setBlockGasLimit_js_1 = require('./actions/test/setBlockGasLimit.js')
Object.defineProperty(exports, 'setBlockGasLimit', {
	enumerable: true,
	get: function () {
		return setBlockGasLimit_js_1.setBlockGasLimit
	},
})
const setBlockTimestampInterval_js_1 = require('./actions/test/setBlockTimestampInterval.js')
Object.defineProperty(exports, 'setBlockTimestampInterval', {
	enumerable: true,
	get: function () {
		return setBlockTimestampInterval_js_1.setBlockTimestampInterval
	},
})
const setCode_js_1 = require('./actions/test/setCode.js')
Object.defineProperty(exports, 'setCode', {
	enumerable: true,
	get: function () {
		return setCode_js_1.setCode
	},
})
const setCoinbase_js_1 = require('./actions/test/setCoinbase.js')
Object.defineProperty(exports, 'setCoinbase', {
	enumerable: true,
	get: function () {
		return setCoinbase_js_1.setCoinbase
	},
})
const setIntervalMining_js_1 = require('./actions/test/setIntervalMining.js')
Object.defineProperty(exports, 'setIntervalMining', {
	enumerable: true,
	get: function () {
		return setIntervalMining_js_1.setIntervalMining
	},
})
const setLoggingEnabled_js_1 = require('./actions/test/setLoggingEnabled.js')
Object.defineProperty(exports, 'setLoggingEnabled', {
	enumerable: true,
	get: function () {
		return setLoggingEnabled_js_1.setLoggingEnabled
	},
})
const setMinGasPrice_js_1 = require('./actions/test/setMinGasPrice.js')
Object.defineProperty(exports, 'setMinGasPrice', {
	enumerable: true,
	get: function () {
		return setMinGasPrice_js_1.setMinGasPrice
	},
})
const setNextBlockBaseFeePerGas_js_1 = require('./actions/test/setNextBlockBaseFeePerGas.js')
Object.defineProperty(exports, 'setNextBlockBaseFeePerGas', {
	enumerable: true,
	get: function () {
		return setNextBlockBaseFeePerGas_js_1.setNextBlockBaseFeePerGas
	},
})
const setNextBlockTimestamp_js_1 = require('./actions/test/setNextBlockTimestamp.js')
Object.defineProperty(exports, 'setNextBlockTimestamp', {
	enumerable: true,
	get: function () {
		return setNextBlockTimestamp_js_1.setNextBlockTimestamp
	},
})
const setNonce_js_1 = require('./actions/test/setNonce.js')
Object.defineProperty(exports, 'setNonce', {
	enumerable: true,
	get: function () {
		return setNonce_js_1.setNonce
	},
})
const setStorageAt_js_1 = require('./actions/test/setStorageAt.js')
Object.defineProperty(exports, 'setStorageAt', {
	enumerable: true,
	get: function () {
		return setStorageAt_js_1.setStorageAt
	},
})
const snapshot_js_1 = require('./actions/test/snapshot.js')
Object.defineProperty(exports, 'snapshot', {
	enumerable: true,
	get: function () {
		return snapshot_js_1.snapshot
	},
})
const setRpcUrl_js_1 = require('./actions/test/setRpcUrl.js')
Object.defineProperty(exports, 'setRpcUrl', {
	enumerable: true,
	get: function () {
		return setRpcUrl_js_1.setRpcUrl
	},
})
const stopImpersonatingAccount_js_1 = require('./actions/test/stopImpersonatingAccount.js')
Object.defineProperty(exports, 'stopImpersonatingAccount', {
	enumerable: true,
	get: function () {
		return stopImpersonatingAccount_js_1.stopImpersonatingAccount
	},
})
//# sourceMappingURL=test.js.map
