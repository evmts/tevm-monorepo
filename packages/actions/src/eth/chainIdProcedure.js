import { numberToHex } from '@tevm/utils'
import { chainIdHandler } from './chainIdHandler.js'

/**
 * @param {import('@tevm/node').TevmNode} baseClient
 * @returns {import('./EthProcedure.js').EthChainIdJsonRpcProcedure}
 */
export const chainIdProcedure = (baseClient) => async (req) => ({
	...(req.id ? { id: req.id } : {}),
	jsonrpc: '2.0',
	method: req.method,
	// TODO pass in a client instead
	result: await chainIdHandler(baseClient)({}).then(numberToHex),
})
