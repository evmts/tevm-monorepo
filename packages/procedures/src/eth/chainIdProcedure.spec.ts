import { describe, expect, it, beforeEach } from 'bun:test'
import { createBaseClient, type BaseClient } from '@tevm/base-client'
import { chainIdProcedure } from './chainIdProcedure.js'
import type { EthChainIdJsonRpcRequest } from './EthJsonRpcRequest.js'

let client: BaseClient

beforeEach(() => {
	client = createBaseClient()
})

describe('chainIdProcedure', () => {
	it('should return the current chain ID', async () => {
		const request: EthChainIdJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_chainId',
			id: 1,
			params: [],
		}

		const response = await chainIdProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_chainId')
		expect(response.id).toBe(request.id as any)
		expect(response.result).toMatchSnapshot()
	})

	it('should handle requests without an id', async () => {
		const request: EthChainIdJsonRpcRequest = {
			jsonrpc: '2.0',
			method: 'eth_chainId',
			params: [],
		}

		const response = await chainIdProcedure(client)(request)
		expect(response.error).toBeUndefined()
		expect(response.result).toBeDefined()
		expect(response.method).toBe('eth_chainId')
		expect(response.id).toBeUndefined()
		expect(response.result).toMatchSnapshot()
	})
})
