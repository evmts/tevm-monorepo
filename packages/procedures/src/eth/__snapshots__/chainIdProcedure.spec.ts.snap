// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`chainIdProcedure should return the current chain ID 1`] = `"0x384"`;

exports[`chainIdProcedure should handle requests without an id 1`] = `"0x384"`;
